### PR TITLE
v5.x

### PR DESCRIPTION
**Pull request by subeeshcbabu**
* Use `swaggerize-routes^2`
    - Use [swagger-parser](https://github.com/BigstickCarpet/swagger-parser) to validate the Swagger spec.
    - The api is dereferenced (both remote and local $ref are dereferenced) using [swagger-parser](https://github.com/BigstickCarpet/swagger-parser) 
    - Use [JSON schema validator](https://github.com/mafintosh/is-my-json-valid) as the default validator. 
    - Option to set `joischema` to `true` to use [Joi](https://github.com/hapijs/joi) schema validator. Uses [enjoi](https://github.com/tlivings/enjoi) - The json to joi schema converter - to build the validator functions.

* Start the server (or `listen` to port) only on `route` event emitted by the app. Route builder is an async api, now capable of parsing and validating remote $ref in swagger spec, so the app/server need to wait for the `route` event before processing requests.

* es6 changes

### CODE REVIEW

File: lib/expressroutes.js [hunk 1]
```diff
@@ -145,104 +42,55 @@ function makeExpressRoute(router, mountpath, route, securityDefinitions) {
...
-function buildNotAllowedMiddleware(methods) {
-    return function (req, res, next) {
-        if (methods.indexOf(req.method.toLowerCase()) === -1) {
-            res.set('Allow', methods.join(', ').toUpperCase());
-            res.sendStatus(405).end();
-        }
-        next();
-    };
-}
+const buildNotAllowedMiddleware = methods => (req, res, next) => {
```
Comment (shaunwarman) [inline lib/expressroutes.js]: Are you setting `methods` and `buildNotAllowedMiddleware ` here?

File: lib/index.js [hunk 2]
```diff
@@ -1,56 +1,45 @@
...
  * @returns {onmount}
  */
-function mount(app, options) {
+const mount = (app, options) => {
 
     return function onmount(parent) {
-        var settings;
-
+        let settings;
+        const routeBuilder = Routes(Object.assign({ }, options));
```
Comment (shaunwarman) [inline lib/index.js]: Extra space

File: lib/index.js [hunk 3]
```diff
@@ -1,56 +1,45 @@
...
+        const errorEvnt = parent.emit.bind(parent, 'error');
```
Comment (shaunwarman) [inline lib/index.js]: Name meant to be shortened?

File: lib/index.js [hunk 4]
```diff
@@ -62,37 +51,29 @@ function mount(app, options) {
             'views': null,
             'view cache': false,
             'view engine': false
-        }).forEach(function (option) {
-            parent.set(option, settings[option]);
-        });
+        }).forEach(option =>  parent.set(option, settings[option]));
```
Comment (shaunwarman) [inline lib/index.js]: Extra space here

File: lib/middlewares/validator.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+const makeValidator = ({ spec, validate }) => (req, res, next) => {
+    let accessor;
+    let inputValue;
+
+    accessor = valueAccessor(spec);
+    inputValue = accessor.get(req, spec.name);
+    let { status, errors, value } = validate(inputValue);
+
+    if (!status && errors && errors.length > 0) {
+        let errorObj = new Error('Validation failed for ' + spec.name);
```
Comment (shaunwarman) [inline lib/middlewares/validator.js]: could use template literals throughout

Comment (subeeshcbabu) [inline lib/middlewares/validator.js]: sure

File: lib/index.js (hunk 4 above)
Comment (subeeshcbabu) [inline lib/index.js]: on it

File: lib/index.js (hunk 3 above)
Comment (subeeshcbabu) [inline lib/index.js]: yeah

File: lib/index.js (hunk 2 above)
Comment (subeeshcbabu) [inline lib/index.js]: on it

File: .jshintrc [hunk 6]
```diff
@@ -1,69 +0,0 @@
-{
```
Comment (gabrielcsapo) [inline .jshintrc]: 👍

File: lib/expressroutes.js [hunk 7]
```diff
@@ -1,141 +1,38 @@
...
 /**
  * Builds a complete path for route usage from the mountpath and the path
  * @param mountpath
  * @param path
  * @return complete route path
  */
-function buildRoutePath(mountpath, path) {
-    return mountpath + utils.prefix(path.replace(/{([^}]+)}/g, ':$1'), '/');
-}
+const buildRoutePath = (mountpath, path) => mountpath + Utils.prefix(path.replace(/{([^}]+)}/g, ':$1'), '/');
```
Comment (gabrielcsapo) [inline lib/expressroutes.js]: 👍

**Comment by gabrielcsapo** [comment]
LGTM @subeeshcbabu 👍

**Comment by gonenduk** [comment]
Can't wait for this release. The previous validations were not good (I even had to fix some myself. Made a pull request - but now I understand why it is not longer relevant).
Please release it as soon as you can. Thanks!

**Comment by gonenduk** [comment]
Any update on release date?

**Comment by gabrielcsapo** [comment]
👍

**Comment by shaunwarman** [comment]
👍 🍾

**Comment by gonenduk** [comment]
any updates on release date?

**Comment by tlivings** [comment]
@subeeshcbabu @shaunwarman you guys want to complete merging this into master and publish it?

**Comment by gonenduk** [comment]
Would be great if you publish the new version to npm

**Comment by gabrielcsapo** [comment]
@subeeshcbabu is working on that this weekend 😄

**Comment by gonenduk** [comment]
Can't wait to try it :)

**Comment by gonenduk** [comment]
any updates?

**Comment by subeeshcbabu-zz** [comment]
Published an alpha version `swaggerize-express@5.0.0-alpha.1`.

Will publish the rc and `5.0.0` soon. Would be awesome if anyone could try this alpha and provide feedback.

**Comment by gonenduk** [comment]
I have an issue with the alpha. Where do I provide feedback? Create a new issue?

**Comment by subeeshcbabu-zz** [comment]
Thanks for trying out the alpha. Yeah, please create an issue with all the details.

**Comment by gonenduk** [comment]
OK, everything is working SUPER well (for me) with the new version.
Any estimates for the release date (with the bug fix)?

**Comment by subeeshcbabu-zz** [comment]
@gonenduk Thanks for testing this out. I will have to publish a version of the `routes` module first and then this one.

I don't see an issue with publishing modules because this is a major version. You can expect something by end of this week.

**Comment by gonenduk** [comment]
Can you also check the PR I did in the routes package (for version 2.x)?
[defaulthandler in options](https://github.com/krakenjs/swaggerize-routes/pull/92)

It's not the same defaulthandler of version 1.x which was quite useless. It's actually a fallback handler in case other handlers are not defined. It's useful to use this to report "501 not implemented" or to create another behavior which is common to all.

**Pull request merged**